pypesto/visualize/waterfall.py

```diff
@@ -0,0 +1,74 @@
+import matplotlib.pyplot as plt
+from matplotlib.ticker import MaxNLocator
+import numpy as np
+from .clust_color import assigncolor
+
+
+def waterfall(result, ax=None):
+
+    """
+    Plot waterfall plot.
+
+    Parameters
+    ----------
+
+    result: pypesto.Result
+        Optimization result obtained by 'optimize.py'
+
+    ax: matplotlib.Axes, optional
+        Axes object to use.
+
+    Returns
+    -------
+
+    ax: matplotlib.Axes
+        The plot axes.
+    """
+
+    result_fval = result.optimize_result.get_for_key('fval')
+    return waterfall_lowlevel(result_fval, ax)
+
+
+def waterfall_lowlevel(result_fval, ax=None):
+
+    """
+    Plot waterfall plot using list of cost function values.
+
+    Parameters
+    ----------
+
+    result_fval: list of cost function values
```

Comment:
We assume here that they are ordered already, right? Should be mentioned or just sort them here.

Reply:
valid point. currently, the multistart results are always automatically sorted after the optimization. i think we should keep doing that, but maybe remember in the result object the index of the optimizer run.

Reply:
Not sure I would appreciate the automated sorting in cases where i provide starting points myself.

Reply:
maybe describe like "numeric list, ordered by 'sorted'"?

Reply:
true, so maybe i delete sorting by default. in the result extraction functions we can maybe add a flag sorted=True, which then does the sorting on a copy list on-the-fly.

Reply:
I mean not now but later, so important for here @DantongWang is just that you do not assume the list to be sorted, and do not do sorting yourself, right?

Reply:
@yannikschaelte sounds good

@DantongWang I think I'd prefer just sorting them

Reply:
@dweindl so what do you mean with "not sure ... see how trajectories diverge", then you would want the sorting routines also to not automatically sort but present the results in the given order?

Reply:
well, sorting routines should sort.. :)

By default I would expect to obtain the optimization results in the order of provided/drawn starting points (In case of parallelization this might become non-trivial for the latter). For convenience, there should additionally be support for sorting by final objective (as implemented, but this should not be the default in my opinion).

Reply:
@yannikschaelte by "do not do sorting yourself" you mean that the list is sorted in the extraction function? And a question: would it still be called a "waterfall plot" if plotted without sorting the cost function values?

Reply:
"do not assume the list to be sorted, and ~do not~ do sorting yourself", i think ;)

Reply:
ups, meant "visualization routines should not automatically sort"

Reply:
but on second thoughts, an unsorted list probably makes not much sense to visualize, so maybe better sort the list.

Reply:
sorry, I did not get it so well. So the visualization routine does not 'automatically' sort the list, but the list can be sorted 'manually' by a flag?

Reply:
> ups, meant "visualization routines should not automatically sort"

since the function is called "waterfall" and not "scatter", i think it should sort automatically.

Reply:
@dweindl Agreed, so I think we can simply move the sorting routine from 'optimization' to 'plotwaterfall'.

Reply:
concerning plotting: should work now (it's sorted in waterfall), but I want to raise awareness that one has to make sure the same coloring is applied in the diverse plotting functions).

atm, in optimizer(), the results are still sorted, but that will be dealt with separately.